File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第9章小明的字符串.py
#小明的字符串  lanqiaoOJ 1203 

N=1000005
Next = [0]*N
def getNext(p):                   #计算Next[1]~Next[plen]
    for i in range(1,len(p)):
        j = Next[i];              #j的后移：j指向前缀阴影w的后一个字符
        while j>0 and p[i] != p[j]:      #阴影的后一个字符不相同
            j = Next[j]              #更新j
        if p[i]==p[j]:   Next[i+1] = j+1
        else:             Next[i+1] = 0
def kmp(s,p):
    ans = 0
    j = 0
    for i in range(0,len(s)):                                    
        while j>0 and s[i]!=p[j]:    #失配了。注意j==0是情况(1)
             j=Next[j]               #j滑动到Next[j]位置
        if s[i]==p[j]:               #当前位置的字符匹配，继续
            j+=1
            ans = max(ans,j)         #去掉这一句，本代码就是最基本的KMP
        if j == len(p): return ans   #j到了P的末尾，找到了一个匹配                       
    return ans                       #返回p在s中出现的最长前缀
s=input()
t=input()  
getNext(t)
print(kmp(s,t))

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第9章小明的字符串.java
//小明的字符串  lanqiaoOJ 1203

import java.util.Scanner;

public class Main {
//https://www.lanqiao.cn/courses/5194/learning/?id=497974
    public static void main(String[] args) {
        Scanner scanner=new Scanner(System.in);
        String s=scanner.next();
        String t=scanner.next();
        getNext(t);
        KMP(s, t);
        System.out.println(max);
    }
    static int next[]=new int [100005];
    static int max=0;
    public static void getNext(String a) {
        int len=a.length();
        int k=-1;
        int j=0;
        next[0]=-1;
        while(j<len-1){
            if(k==-1 || a.charAt(j)==a.charAt(k)){
                j++;
                k++;
                next[j]=k;
                max=Math.max(max, next[j]);
            }else{
                k=next[k];
                max=Math.max(max, k);
            }
        }
    }
    
    static int KMP(String a,String b){
        int alen=a.length();
        int blen=b.length();
        int i=0;
        int j=-1;
        while(i<alen && j<blen){
            if(j==-1 || a.charAt(i)==b.charAt(j)){
                i++;j++;
            }else
                j=next[j];
            max=Math.max(max, j);
        }
        if(j==blen){
            return i-j;
        }else
            return -1;
    }
}